### PR TITLE
Filter out page/sort/size as field names

### LINKED ISSUE #14199: Entities broken with pagination + criteria filter
##### **Overview of the issue**

Can't list entity with using pagination + criteria filter

beer_entity

```
2021-03-06 19:15:16.370  WARN 18032 --- [  XNIO-1 task-2] o.z.problem.spring.common.AdviceTraits   : Bad Request: org.springframework.validation.BeanPropertyBindingResult: 1 errors
Field error in object 'productCriteria' on field 'size': rejected value [20]; codes [typeMismatch.productCriteria.size,typeMismatch.size,typeMismatch.io.github.pascalgrimaud.beer.service.criteria.ProductCriteria$SizeFilter,typeMismatch]; arguments [org.springframework.context.support.DefaultMessageSourceResolvable: codes [productCriteria.size,size]; arguments []; default message [size]]; default message [Failed to convert property value of type 'java.lang.String' to required type 'io.github.pascalgrimaud.beer.service.criteria.ProductCriteria$SizeFilter' for property 'size'; nested exception is java.lang.IllegalStateException: Cannot convert value of type 'java.lang.String' to required type 'io.github.pascalgrimaud.beer.service.criteria.ProductCriteria$SizeFilter' for property 'size': no matching editors or conversion strategy found]
```

##### **Suggest a Fix**

It's probably related to ENUM, but not sure

##### **JHipster Version(s)**

```
beer@0.0.0 /home/pgrimaud/tmp/12-beer
└── generator-jhipster@7.0.0-beta.1 

```

##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**

<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "authenticationType": "jwt",
    "cacheProvider": "ehcache",
    "clientFramework": "vue",
    "serverPort": "8080",
    "serviceDiscoveryType": false,
    "skipUserManagement": false,
    "withAdminUi": true,
    "baseName": "beer",
    "buildTool": "maven",
    "databaseType": "sql",
    "devDatabaseType": "postgresql",
    "enableHibernateCache": true,
    "enableSwaggerCodegen": false,
    "enableTranslation": true,
    "jhiPrefix": "jhi",
    "languages": ["en"],
    "messageBroker": false,
    "prodDatabaseType": "postgresql",
    "searchEngine": false,
    "skipClient": false,
    "testFrameworks": ["cypress"],
    "websocket": false,
    "applicationType": "monolith",
    "packageName": "io.github.pascalgrimaud.beer",
    "packageFolder": "io/github/pascalgrimaud/beer",
    "jhipsterVersion": "7.0.0-beta.1",
    "skipServer": false,
    "clientPackageManager": "npm",
    "dtoSuffix": "DTO",
    "entitySuffix": "",
    "reactive": false,
    "clientTheme": "none",
    "clientThemeVariant": "",
    "entities": ["Product"],
    "skipCheckLengthOfIdentifier": false,
    "skipFakeData": false,
    "blueprints": [],
    "otherModules": [],
    "pages": [],
    "nativeLanguage": "en",
    "creationTimestamp": 1615053884446,
    "jwtSecretKey": "YourJWTSecretKeyWasReplacedByThisMeaninglessTextByTheJHipsterInfoCommandForObviousSecurityReasons",
    "lastLiquibaseTimestamp": 1615053944000
  }
}

</pre>
</details>

##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

<details>
<summary>JDL entity definitions</summary>

<pre>
entity Product {
  name String required
  description String
  price BigDecimal required min(0)
  size Size required
  image ImageBlob
}
enum Size {
  S,
  M,
  L,
  XL,
  XXL
}

dto Product with mapstruct
paginate Product with pagination
service Product with serviceClass
filter Product

</pre>
</details>

My JDL

```
application {
  config {
    baseName beer
    applicationType monolith
    packageName io.github.pascalgrimaud.beer
    authenticationType jwt
    devDatabaseType postgresql
    prodDatabaseType postgresql
    buildTool maven
    clientFramework vue
    testFrameworks [cypress]
  }
  entities *
}

entity Product {
  name String required
  description String
  price BigDecimal required min(0)
  size Size required
  image ImageBlob
}
enum Size {
  S, M, L, XL, XXL
}

dto Product with mapstruct
service Product with serviceClass
paginate Product with pagination
filter Product
```

##### **Environment and Tools**

openjdk version "11.0.10" 2021-01-19
OpenJDK Runtime Environment (build 11.0.10+9-Ubuntu-0ubuntu1.18.04)
OpenJDK 64-Bit Server VM (build 11.0.10+9-Ubuntu-0ubuntu1.18.04, mixed mode, sharing)

git version 2.30.1

node: v14.15.4

npm: 6.14.10

yeoman: 3.1.1

Docker version 19.03.8, build afacb8b7f0

docker-compose version 1.27.3, build 4092ae5d

INFO! Congratulations, JHipster execution is complete!

### PR DESCRIPTION
**Pull request by gzsombor**
Fix #14199 

---

Please make sure the below checklist is followed for Pull Requests.

- [x] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [x] Tests are added where necessary
- [x] [jhipster-online](https://github.com/jhipster/jhipster-online) was updated if necessary
- [x] Documentation is added/updated where necessary
- [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (bellow reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

File: jdl/validators/jdl-with-application-validator.js [hunk 1]
```diff
@@ -104,12 +105,20 @@ function createValidator(jdlObject, logger = console) {
...
+      jdlApplication.getConfigurationOptionValue('reactive') === false;
     Object.keys(jdlFields).forEach(fieldName => {
       const jdlField = jdlFields[fieldName];
       validator.validate(jdlField);
       if (isReservedFieldName(jdlField.name)) {
         logger.warn(`The name '${jdlField.name}' is a reserved keyword, so it will be prefixed with the value of 'jhiPrefix'.`);
       }
+      if (filtering && isReservedPaginationWords(jdlField.name)) {
+        throw new Error(
+          `Field name '${fieldName}'found in ${entityName} is a reserved keyword, as it is used by Spring for pagination in the URL.`
```
Comment (pascalgrimaud) [inline jdl/validators/jdl-with-application-validator.js:119]:
```suggestion
          `Field name '${fieldName}' found in ${entityName} is a reserved keyword, as it is used by Spring for pagination in the URL.`
```

File: jdl/validators/jdl-without-application-validator.js [hunk 2]
```diff
@@ -91,12 +92,18 @@ function createValidator(jdlObject, applicationSettings = {}, logger = console)
...
+    const filtering = applicationSettings.databaseType === 'sql';
     Object.keys(jdlFields).forEach(fieldName => {
       const jdlField = jdlFields[fieldName];
       validator.validate(jdlField);
       if (isReservedFieldName(jdlField.name)) {
         logger.warn(`The name '${jdlField.name}' is a reserved keyword, so it will be prefixed with the value of 'jhiPrefix'.`);
       }
+      if (filtering && isReservedPaginationWords(jdlField.name)) {
+        throw new Error(
+          `Field name '${fieldName}'found in ${entityName} is a reserved keyword, as it is used by Spring for pagination in the URL.`
```
Comment (pascalgrimaud) [inline jdl/validators/jdl-without-application-validator.js:104]:
```suggestion
          `Field name '${fieldName}' found in ${entityName} is a reserved keyword, as it is used by Spring for pagination in the URL.`
```

**Review by pascalgrimaud — CHANGES_REQUESTED** [review]

**Review by pascalgrimaud — APPROVED** [review]

**Comment by pascalgrimaud** [comment]
Thanks so much for your quick fix @gzsombor

**Comment by gzsombor** [comment]
Thanks and welcome!

**Comment by pascalgrimaud** [comment]
@gzsombor : don't forget your bounty :-)

**Comment by gzsombor** [comment]
Thanks, bounty claimed: https://opencollective.com/generator-jhipster/expenses/35782 :smiley:

**Comment by pascalgrimaud** [comment]
@gzsombor : approved !

**Pull request merged**